### PR TITLE
feat(linter): optimize lint performance. resolves #5210

### LINKED ISSUE #5210: enforce-module-boundaries is extremely slow even with default configuration
We're migrating our monorepo to eslint (~50 apps, 360 libs, nothing configured in nx.json (empty tags arrays)). After some time of debugging and reducing received eslint config we found that enabling `enforce-module-boundaries` increases lint time by ~7 times

## Current Behavior

Adding this rule to our eslint config
```json
{
  "files": ["*.ts", "*.js"],
  "rules": {
    "@nrwl/nx/enforce-module-boundaries": [
      "error",
      {
        "enforceBuildableLibDependency": true,
        "allow": [],
        "depConstraints": [
          {
            "sourceTag": "*",
            "onlyDependOnLibsWithTags": ["*"]
          }
        ]
      }
    ]
  }
}
```
increases lint time (for one of our projects) from 60s to 410s. We have no tags in nx config.

## Expected Behavior

`enforce-module-boundaries` should not be that slow

## Steps to Reproduce

Our code is private. I can use screen share (drow#4335 in angular discord) or provide any additional debug information.

### Failure Logs

eslint timings debug
```
Rule                                       |  Time (ms) | Relative
:------------------------------------------|-----------:|--------:
@nrwl/nx/enforce-module-boundaries         | 286411.805 |    93.7%
import/no-deprecated                       |  15667.608 |     5.1%
no-redeclare                               |    751.399 |     0.2%
jsdoc/check-alignment                      |    433.630 |     0.1%
@typescript-eslint/naming-convention       |    378.335 |     0.1%
jsdoc/newline-after-description            |    201.155 |     0.1%
max-len                                    |    162.705 |     0.1%
object-shorthand                           |    133.897 |     0.0%
id-match                                   |     92.267 |     0.0%
@typescript-eslint/type-annotation-spacing |     90.929 |     0.0%
```

eslint config (almost nothing in project's one)
https://pastebin.com/ekKLVupp

### Environment

```
  Node : 12.18.1
  OS   : win32 x64
  yarn : 1.22.10

  nx : Not Found
  @nrwl/angular : 11.6.1
  @nrwl/cli : 11.6.1
  @nrwl/cypress : 11.6.1
  @nrwl/devkit : 11.6.1
  @nrwl/eslint-plugin-nx : 11.6.1
  @nrwl/express : Not Found
  @nrwl/jest : 11.6.1
  @nrwl/linter : 11.6.1
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : 11.6.1
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.6.1
  @nrwl/web : Not Found
  @nrwl/workspace : 11.6.1
  @nrwl/storybook : Not Found
  @nrwl/gatsby : Not Found
  typescript : 4.1.3
```

### PR DESCRIPTION
**Pull request by meeroslav**
## Current Behavior
Running `enforce-module-boundaries` rule in linter is slow. This rule can take up to 90% of the entire linting duration.
Most of this time is spent on finding the parent source project and target project from import.

This is an example of such report:
```
Rule                               | Time (ms) | Relative
:----------------------------------|----------:|--------:
@nrwl/nx/enforce-module-boundaries |    45.131 |    54.0%
@typescript-eslint/no-unused-vars  |     5.472 |     6.5%
react/no-direct-mutation-state     |     4.385 |     5.2%
no-restricted-globals              |     1.941 |     2.3%
react/jsx-no-comment-textnodes     |     1.567 |     1.9%
react/no-typos                     |     1.539 |     1.8%
react-hooks/rules-of-hooks         |     1.532 |     1.8%
react/require-render-return        |     1.416 |     1.7%
no-octal-escape                    |     1.385 |     1.7%
no-global-assign                   |     1.359 |     1.6%
```

## Expected Behavior
The duration of `enforce-module-boundaries` should not be significantly higher than that of other lint rules.

## Related Issue(s)

Fixes #5210

### CODE REVIEW

**Comment by vsavkin** [comment]
It looks. Once the CI is green, you can merge it.

File: packages/workspace/src/utils/graph-utils.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import {
 import { isWorkspaceProject } from '../core/project-graph/operators';
 
 interface Reach {
-  graph: ProjectGraph;
+  graph: ProjectGraph<any, any>;
```
Comment (vsavkin) [inline packages/workspace/src/utils/graph-utils.ts:8]: Why is this necessary? Why don't default type values work?

File: packages/workspace/src/utils/graph-utils.ts [hunk 2]
```diff
@@ -65,10 +65,10 @@ function buildMatrix(graph: ProjectGraph) {
 }
 
 export function getPath(
-  graph: ProjectGraph,
+  graph: ProjectGraph<any, any>,
   sourceProjectName: string,
   targetProjectName: string
-): Array<ProjectGraphNode> {
+): Array<ProjectGraphNode<any, any>> {
```
Comment (vsavkin) [inline packages/workspace/src/utils/graph-utils.ts:71]: same question here

File: packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts [hunk 3]
```diff
@@ -129,7 +130,9 @@ export default createESLintRule<Options, MessageIds>({
     }
 
     const npmScope = (global as any).npmScope;
-    const projectGraph = (global as any).projectGraph as ProjectGraph;
+    const projectGraph = mapProjectGraphFiles(
```
Comment (vsavkin) [inline packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts:133]: I think remapping will happen multiple times, cause the object will remain when is used by the IDE for linting

File: packages/workspace/src/utils/runtime-lint-utils.ts [hunk 4]
```diff
@@ -172,11 +167,45 @@ export function getSourceFilePath(sourceFileName: string, projectPath: string) {
...
   );
 }
+
+export function mapProjectGraphFiles<T>(
+  projectGraph: ProjectGraph<T>
+): ProjectGraph<T, Record<string, FileData>> {
+  if (!projectGraph) {
+    return;
+  }
+  const nodes = Object.entries(projectGraph.nodes).reduce(
```
Comment (vsavkin) [inline packages/workspace/src/utils/runtime-lint-utils.ts:187]: Don't use reduce. Hard to read. A nested for loop is a lot better.

**Review by vsavkin — CHANGES_REQUESTED** [review]
Two points:

1. We can only change the interface of ProjectGraph in a backwards compatible way. 
2. I think it's confusing to have an OR type where we use one value for linting and the other one for other use cases. I'd rather create a separate object (the map) for linting and pass that map to the utils in runtime-lint-utils. So ProjectGraph remains as is, and findSourceProject and findTargetProject take the map as an argument.

File: packages/workspace/src/utils/graph-utils.ts (hunk 1 above)
Comment (meeroslav) [reply packages/workspace/src/utils/graph-utils.ts:8]: Default second generic type falls back to `FileData[]` which is then not type-compatible with `Record<string, FileData>`.

File: packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts (hunk 3 above)
Comment (meeroslav) [reply packages/eslint-plugin-nx/src/rules/enforce-module-boundaries.ts:133]: You are right, I will move it to the branch above.

File: packages/workspace/src/utilities/fileutils.ts [hunk 5]
```diff
@@ -47,7 +47,11 @@ export function readJsonFile<T = any>(path: string): T {
 }
 
 export function parseJsonWithComments<T = any>(content: string): T {
-  return JSON.parse(stripJsonComments(content));
+  try {
```
[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/utilities/fileutils.ts:50]: Currently it's roughly 5-10% (~35ms to 32ms on nx-incremental-repo single project). It used to be more significant before other optimisations, when we were doing more I/O + parsing.

Based on the benchmark in the official [strip-json-comments](https://github.com/sindresorhus/strip-json-comments) for the average size JSON file (1000 lines) it takes 1-2ms (~800op/s)

[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/utilities/fileutils.ts:50]: It's not a significant increase, but most of JSON files don't have comments, so it makes sense to do it as a corner case rather than a happy flow IMHO.

[the remark this answers is not in the document]
Comment (meeroslav) [reply packages/workspace/src/utilities/fileutils.ts:50]: Here are the direct comparisons where reading + parsing `nx.json` and `nxdeps.json`:

With `strip-json-comments` as a default operation:
```
0.544791        Read nx.json
68.505417      Read nxdeps.json
```

With `strip-json-comments` as a fallback operation:
```
0.188125        Read nx.json
5.084000       Read nxdeps.json
```

To the lint rule itself in runtime it has no longer a big impact, as most of this heavy lifting is done once on the rule creation.

**Comment by vsavkin** [comment]
You need to rebase, which should fix CI, otherwise looks good. Merge it1

**Review by vsavkin — APPROVED** [review]
